spec/milestone_checker_spec.rb

```diff
@@ -0,0 +1,195 @@
+            'milestone' => {
+              'title' => 'Release Day',
+              'html_url' => 'https://wp.com',
+              'due_on' => '2023-06-30T23:59:01Z'
+            },
+            'state' => 'open'
+          }
+
+          date_nine_days_before_due = DateTime.parse('2023-06-21T23:59:01Z')
+
+          allow(@plugin.github).to receive(:pr_json).and_return(pr_json)
+          allow(DateTime).to receive(:now).and_return(date_nine_days_before_due)
+
+          @plugin.check_milestone_due_date(warning_days: 10)
+
+          expected_warning = ["This PR is assigned to the milestone [Release Day](https://wp.com). This milestone is due in less than 10 days.\nPlease make sure to get it merged by then or assign it to a milestone with a later deadline."]
+          expect(@dangerfile.status_report[:warnings]).to eq expected_warning
+        end
+
+        it 'does nothing when a PR has a milestone without a due date' do
+          pr_json = {
+            'milestone' => {
+              'title' => 'Release Day',
+              'html_url' => 'https://wp.com'
+            },
+            'state' => 'open'
+          }
+
+          allow(@plugin.github).to receive(:pr_json).and_return(pr_json)
+
+          @plugin.check_milestone_due_date
+
+          expect(@dangerfile.status_report[:warnings]).to be_empty
+        end
+
+        it 'does nothing when a PR has a milestone but it has already passed the due date' do
+          pr_json = {
+            'milestone' => {
+              'title' => 'Release Day',
+              'html_url' => 'https://wp.com',
+              'due_on' => '2023-06-30T23:59:01Z'
+            },
+            'state' => 'closed'
+          }
+
+          allow(@plugin.github).to receive(:pr_json).and_return(pr_json)
+
+          @plugin.check_milestone_due_date
+
+          expect(@dangerfile.status_report[:warnings]).to be_empty
+        end
+
+        it "reports a warning when asked to do so when a PR doesn't have a milestone set" do
+          allow(@plugin.github).to receive(:pr_json).and_return({})
+
+          @plugin.check_milestone_due_date(if_no_milestone: :warn)
+
+          expected_warning = ['PR is not assigned to a milestone.']
+          expect(@dangerfile.status_report[:warnings]).to eq expected_warning
+        end
+
+        it "reports an error when asked to do so when a PR doesn't have a milestone set" do
+          allow(@plugin.github).to receive(:pr_json).and_return({})
+
+          @plugin.check_milestone_due_date(if_no_milestone: :error)
+
+          expected_error = ['PR is not assigned to a milestone.']
+          expect(@dangerfile.status_report[:errors]).to eq expected_error
+        end
+
+        it "does nothing when asked to do so when a PR doesn't have a milestone set" do
+          allow(@plugin.github).to receive(:pr_json).and_return({})
+
+          @plugin.check_milestone_due_date(if_no_milestone: :none)
+
+          expect(@dangerfile.status_report[:errors]).to be_empty
```

Comment:
I think this might be a false positive test. 

The current implementation is calls

```ruby
        check_milestone_set(fail_on_error: if_no_milestone == :error)
```

meaning that in this case it'll report a warning.

Reply:
Confirmed locally:

<img width="2047" alt="image" src="https://github.com/Automattic/dangermattic/assets/1218433/a68cc8e3-803a-461d-bcfc-618396357dbc">

Reply:
This instance reinforces my belief that we would benefit from consistently checking both errors and warnings.

Maybe we could write a custom expectation or just a util method in `spec_helper` 🤔 Anyways, that's something for another PR.

Reply:
Nice catch! When I introduced the error/warning, the tests also should have been updated to check for both.
I've fixed this here e51cb9a83bf03ca6bb095aae5f42c7129414165c and here d83980810554ce504e6757745a1dbc933525247f.

While I'm at it, I'm going to create a PR on top of this one to use a new expectation util method 👍